src/memory_region.rs

```diff
@@ -52,68 +52,102 @@ impl fmt::Debug for MemoryRegion {
-    access_type: &str,
-    pc: usize, // TODO syscalls don't have this info
-    regions: &[MemoryRegion],
-) -> Result<u64, EbpfError<E>> {
-    let index = match regions.binary_search_by(|probe| probe.addr_vm.cmp(&vm_addr)) {
-        Ok(index) => index,
-        Err(index) => {
-            if index == 0 {
-                return Err(generate_access_violation(
-                    vm_addr,
-                    len,
-                    access_type,
-                    pc,
-                    regions,
-                ));
+    /// Adds a region
+    pub fn add_region(&mut self, region: MemoryRegion) {
+        self.regions.push(region);
+    }
+
+    /// Adds multiple regions
+    pub fn add_regions(&mut self, regions: Vec<MemoryRegion>) {
+        self.regions.extend(regions);
+    }
+
+    /// Call after the last change. Sorts regions by addr_vm for binary search
+    pub fn finalize(&mut self) {
+        self.regions.sort_by(|a, b| a.addr_vm.cmp(&b.addr_vm));
+    }
+
+    /// Given a list of regions translate from virtual machine to host address
+    pub fn translate_addr<E: UserDefinedError>(
+        &self,
+        access_type: AccessType,
+        vm_addr: u64,
+        len: u64,
+    ) -> Result<u64, EbpfError<E>> {
+        let index = match self
+            .regions
+            .binary_search_by(|probe| probe.addr_vm.cmp(&vm_addr))
+        {
+            Ok(index) => index,
+            Err(index) => {
+                if index == 0 {
+                    return Err(self.generate_access_violation(access_type, vm_addr, len));
+                }
+                index - 1
+            }
+        };
+        let region = &self.regions[index];
+        if access_type == AccessType::Load || region.writable {
+            if let Ok(host_addr) = region.vm_to_host::<E>(vm_addr, len as u64) {
+                return Ok(host_addr);
             }
-            index - 1
         }
-    };
+        Err(self.generate_access_violation(access_type, vm_addr, len))
+    }
 
-    if let Ok(host_addr) = regions[index].vm_to_host::<E>(vm_addr, len as u64) {
-        Ok(host_addr)
-    } else {
-        Err(generate_access_violation(
+    /// Helper for translate_addr to generate errors
+    fn generate_access_violation<E: UserDefinedError>(
+        &self,
+        access_type: AccessType,
+        vm_addr: u64,
+        len: u64,
+    ) -> EbpfError<E> {
+        let mut regions_string = "".to_string();
+        if !self.regions.is_empty() {
+            regions_string = "regions:".to_string();
+            for region in self.regions.iter() {
+                regions_string = format!(
+                    "  {} \n{:#x} {:#x} {:#x}",
+                    regions_string, region.addr_host, region.addr_vm, region.len,
+                );
+            }
+        }
+        EbpfError::AccessViolation(
+            0, // Filled out later
+            access_type,
             vm_addr,
             len,
-            access_type,
-            pc,
```

Comment:
The pc where the violation occurred is not longer reported?  You got rid of this because in jit you don't have easy access to this information?  If so, maybe we have some kind of "unknown" value so developers still get this information when using the interpretor

Reply:
It is available in both interpreter and JIT. It just is not passed into `MemoryMapping::translate_addr`, but set after an error occurs in `translate_memory_access` for the interpreter and `emit_address_translation` for the JIT. This saves us one argument in each call to `MemoryMapping::translate_addr`.